File: lib/dev_kits/py/sample_toolkit.py
import lib

# Extend GameState class to add our own methods
class Coordinate(object):
  def __init__(self,x,y):
    self.x = x
    self.y = y

  def __add__(self,coord):
    return Coordinate(self.x + coord.x, self.y + coord.y)

  def __repr__(self):
    return "Coordinate(%d, %d)" % (self.x, self.y)

INVALID_COORDINATE = Coordinate(-1,-1)


class GameStateSample(lib.GameState):
    
  ALL_DIRECTIONS = [Coordinate(0,1), Coordinate(0,-1), Coordinate(1,0), Coordinate(-1,0)]
    
  def at(self,x,y):
    if x < 0 or x >= self.width: return None 
    if y < 0 or y >= self.height: return None 
    return self.board[x][y]

  def is_available(self,x,y):
    return self.at(x, y) == lib.BLANK

  def is_wall(self,x,y):
    return self.at(x, y) == lib.WALL

  def is_player(self,x,y):
    return self.at(x, y) and not self.is_wall(x, y) and not self.is_available(x, y)

  def is_same_player(self,x,y,team):
    return self.at(x, y) == team
    
  def is_valid(self,move):
    # Empty space condition
    if not self.is_available(move.x, move.y): return False
    # Adjacency condition
    return self.is_adjacent(move)

  def is_adjacent(self,move):
    for direction in GameStateSample.ALL_DIRECTIONS:
      adjacent_cell = move + direction
      if self.is_same_player(adjacent_cell.x, adjacent_cell.y, lib.ME): return True
    return False
  
class SampleBotBase(object):
  def get_move(self,game_state):
    """Implement this method in derived sample"""

  def get_possible_moves(self,game_state,position):
    """Provides the list of all moves from one position"""
    position_list = []
    for direction in GameStateSample.ALL_DIRECTIONS:
      move = position + direction
      if game_state.is_valid(move): position_list.append(move)
    return position_list

  def get_all_possible_moves(self,game_state):
    """Provides all the possible moves available for the player"""
    territory_list = self.get_territories(game_state, lib.ME)
    position_list = []
    for position in territory_list:
      position_list.extend(self.get_possible_moves(game_state, position))
    return position_list

  def get_territories(self,game_state,team):
    """Provides the list of all territories for the specified team"""
    territory_list = []
    for i in range(0, game_state.width):
      for j in range(0, game_state.height):
        if game_state.is_same_player(i, j, team): territory_list.append(Coordinate(i, j))
    return territory_list
 

File: lib/dev_kits/rb/sample_toolkit.rb
require 'lib'

# Extend GameState class to add our own methods
module BlobWars
  class GameState
    
    ALL_DIRECTIONS = [Coordinate.new(0,1), Coordinate.new(0,-1), Coordinate.new(1,0), Coordinate.new(-1,0)]
    
    def at(x,y)
      # Make sure we are in border
      return nil if x < 0 || x >= @width
      return nil if y < 0 || y >= @height

      @board[x][y]
    end

    # Cells status
    def available?(x,y)
      at(x, y) == BLANK
    end

    def wall?(x,y)
      at(x, y) == WALL
    end

    def player?(x,y)
      at(x, y) != nil && !wall?(x, y) && !available?(x, y)
    end

    def same_player?(x,y,team)
      at(x, y) == team
    end
    
    def valid?(move)
      # Empty space condition
      return false if !available?(move.x, move.y)

      # Adjacency condition
      adjacency?(game_state, move)
    end

    def adjacency?(move)
      ALL_DIRECTIONS.each do |direction|
        adjacent_cell = move + direction
        true if same_player?(adjacent_cell.x, adjacent_cell.y, ME)
      end
      false
    end
  end
end

class Coordinate
  def initialize(x, y)
    @x = x
    @y = y
  end
  
  def +(coord)
    return Coordinate.new(x + coord.x, y + coord.y)
  end

  attr_accessor :x, :y
end

class SampleBotBase
  
  def get_move(game_state)
    # Implement this method in derived sample
  end

  # Provides the list of all moves from one position
  def get_possible_moves(game_state, position)
    position_list = []
    game_state.ALL_DIRECTIONS.each do |direction|
      move = position + direction
      if game_state.valid?(move)
        position_list << move
      end
    end
    position_list
  end

  # Provides all the possible moves available for the player
  def get_all_possible_moves(game_state)
    territory_list = get_territories(game_state)
    position_list = []
    territory_list.each do |position|
      position_list.concat(get_possible_moves(game_state, position))
    end
    position_list
  end

  # Provides the list of all territories for the specified team
  def get_territories(game_state)
    territory_list = []
    (0...game_state.width).each do |i| 
      (0...game_state.height).each do |j|
        if game_state.same_player?(i, j)
          territory_list << Coordinate.new(i, j) 
        end
      end
    end
    territory_list
  end

end
